### PR TITLE
fix: answer an out-of-range cast per dialect, and translate MySQL's MOD operator

### LINKED ISSUE #389: A cast of a value past the integer range returns the most negative integer instead of the dialect's answer
Found on c81a0eb while probing S6 (dialect translation) with differential testing against equivalent SQLite queries.

## Phenomenon

A cast of a number too large for a 64-bit integer returns the most negative one instead of being rejected: `CAST(1e30 AS SIGNED)` in the MySQL dialect, `(1e30)::bigint` in PostgreSQL, and `CAST(1e30 AS INT64)` and `SAFE_CAST(1e30 AS INT64)` in GoogleSQL all answer `-9223372036854775808`, with no error and nothing to say the value did not fit. A positive value comes back negative, so a comparison against it is inverted and an aggregate over a column holding one is wrong by 2^64; a query that casts to validate its input reports success on exactly the rows it was written to reject, which is the failure the cast helpers exist to prevent. The same happens for a string source (`CAST('99999999999999999999' AS SIGNED)`) and for infinity (`CAST(1e308*10 AS SIGNED)`).

## Reproduction

```go
package repro

import (
	"context"
	"database/sql"
	"strings"
	"testing"

	"github.com/nao1215/filesql"
	"github.com/nao1215/filesql/dialect"
)

func TestCastOutOfRangeIsSilent(t *testing.T) {
	for _, tc := range []struct {
		d     dialect.Dialect
		query string
	}{
		{dialect.MySQL, "SELECT CAST(1e30 AS SIGNED)"},
		{dialect.MySQL, "SELECT CAST('99999999999999999999' AS SIGNED)"},
		{dialect.PostgreSQL, "SELECT (1e30)::bigint"},
		{dialect.GoogleSQL, "SELECT CAST(1e30 AS INT64)"},
		{dialect.GoogleSQL, "SELECT SAFE_CAST(1e30 AS INT64)"},
	} {
		ctx := context.Background()
		built, err := filesql.NewBuilder().
			AddReader(strings.NewReader("a,b\n7,2\n"), "t", filesql.FileTypeCSV).
			WithDialect(tc.d).Build(ctx)
		if err != nil {
			t.Fatal(err)
		}
		db, err := built.Open(ctx)
		if err != nil {
			t.Fatal(err)
		}
		var got sql.NullString
		err = db.QueryRow(tc.query).Scan(&got)
		t.Logf("%-10s %-42s -> %v (valid %v) err=%v", tc.d, tc.query, got.String, got.Valid, err)
		db.Close()
	}
}
```

```
    dialect_test.go:43: mysql      SELECT CAST(1e30 AS SIGNED)                -> -9223372036854775808 (valid true) err=<nil>
    dialect_test.go:43: mysql      SELECT CAST('99999999999999999999' AS SIGNED) -> -9223372036854775808 (valid true) err=<nil>
    dialect_test.go:43: postgresql SELECT (1e30)::bigint                      -> -9223372036854775808 (valid true) err=<nil>
    dialect_test.go:43: googlesql  SELECT CAST(1e30 AS INT64)                 -> -9223372036854775808 (valid true) err=<nil>
    dialect_test.go:43: googlesql  SELECT SAFE_CAST(1e30 AS INT64)            -> -9223372036854775808 (valid true) err=<nil>
```

Every one of the five returns the most negative int64 for a value that is positive and far outside the range of the target type.

## Expected behavior

Each dialect has an answer for a value the integer type cannot hold, and none of them is this one. MySQL clamps to the bound of the type and warns, so `CAST(1e30 AS SIGNED)` is `9223372036854775807`. PostgreSQL raises `bigint out of range`. GoogleSQL raises for `CAST` and returns NULL for `SAFE_CAST`, which is what `SAFE_CAST` is for. The package documents the rule it is breaking in `dialect/doc.go`: SQLite's own CAST "coerces a value the target type cannot represent instead of rejecting it. Each dialect therefore rewrites CAST (and PostgreSQL's `::`, and GoogleSQL's `SAFE_CAST`) into a helper that converts with that dialect's rules and returns ErrInvalidCast for a value it cannot represent." The helper is reached and then produces a coerced value of its own.

Beyond being wrong, the value is not defined: the Go specification says a conversion of a floating-point value the integer type cannot represent succeeds with an implementation-dependent result, so the same query need not answer the same thing on the platforms this project builds for.

## Cause (guess)

This is a guess: `roundForDialect` in `dialect/cast.go` ends in `int64(math.Round(f))` (and `int64(math.RoundToEven(f))` for PostgreSQL) with nothing between the rounding and the conversion, so a value outside the int64 range, and an infinity, reach the conversion the Go spec leaves implementation-dependent. `castToInt` calls it for a float64 source, for a string that parses as a float, and for MySQL's numeric-prefix coercion, which is why all three paths show it. The fix is to check the rounded float against the range before converting — a float64 holds no int64 exactly above 2^63, so the bound to compare against is `math.MaxInt64` as a float, and NaN has to be handled too — and then to answer per dialect: clamp for MySQL, `ErrInvalidCast` for PostgreSQL and GoogleSQL, which `SAFE_CAST` already turns into NULL.

## Tests to add

- In the dialect cast tests, a table over the three dialects asserting the boundary: `9223372036854775807` and `-9223372036854775808` cast to an integer type keep their value, and `1e30`, `-1e30`, `1e308*10`, `-1e308*10` and `'99999999999999999999'` give the dialect's answer — the clamp for MySQL, `ErrInvalidCast` for PostgreSQL, an error for GoogleSQL's `CAST` and NULL for its `SAFE_CAST`.
- The sibling case for the other direction of the same mistake: a value just inside the range must not be rejected, so `9.2e18` and `-9.2e18` still convert, and `CAST(9223372036854775807 AS SIGNED)` is unchanged.
- A case for NaN, which is neither in range nor out of it: `CAST('nan' AS SIGNED)` in each dialect, so the range check does not let it through by comparing false against both bounds.
- The float and decimal targets deserve the same look while this is being fixed: `CAST('1e400' AS DOUBLE)` is an infinity that SQLite has no way to store, and a test should pin whatever answer is chosen for it.

### LINKED ISSUE #390: MySQL's MOD operator reaches SQLite unchanged and fails as a syntax error
Found on c81a0eb while probing S6 (dialect translation) with differential testing against equivalent SQLite queries.

## Phenomenon

MySQL's `MOD` operator is passed to SQLite unchanged and comes back as a syntax error naming a token the caller did not write: `SELECT 7 MOD 2` fails with `near "2": syntax error` and `SELECT * FROM t WHERE a MOD b = 1` with `near "MOD": syntax error`. `MOD` is MySQL's spelling of the remainder operator and has an exact SQLite equivalent in `%`, and its sibling `DIV` is already rewritten, so a query using one of the pair works and the same query using the other fails on the engine message this package exists to keep callers away from. `MOD(a, b)`, the function spelling, works, which makes the failure look like a parser problem in the caller's query rather than a gap here.

## Reproduction

```go
package repro

import (
	"context"
	"database/sql"
	"strings"
	"testing"

	"github.com/nao1215/filesql"
	"github.com/nao1215/filesql/dialect"
)

func TestMySQLModOperator(t *testing.T) {
	ctx := context.Background()
	built, err := filesql.NewBuilder().
		AddReader(strings.NewReader("a,b\n7,2\n"), "t", filesql.FileTypeCSV).
		WithDialect(dialect.MySQL).Build(ctx)
	if err != nil {
		t.Fatal(err)
	}
	db, err := built.Open(ctx)
	if err != nil {
		t.Fatal(err)
	}
	defer db.Close()

	for _, query := range []string{
		"SELECT 7 MOD 2",
		"SELECT a MOD b FROM t",
		"SELECT 7 DIV 2",
		"SELECT MOD(7, 2)",
	} {
		translated, terr := dialect.Translate(dialect.MySQL, query)
		var got sql.NullString
		err := db.QueryRow(query).Scan(&got)
		t.Logf("%-24s -> %-58s run: %v err=%v (translate err %v)", query, translated, got.String, err, terr)
	}
}
```

```
    dialect_test.go:60: SELECT 7 MOD 2           -> SELECT 7 MOD 2                                             run:  err=SQL logic error: near "2": syntax error (1) (translate err <nil>)
    dialect_test.go:60: SELECT a MOD b FROM t    -> SELECT a MOD b FROM t                                      run:  err=SQL logic error: near "b": syntax error (1) (translate err <nil>)
    dialect_test.go:60: SELECT 7 DIV 2           -> SELECT CAST(mysql_divide(7, 2) AS INTEGER) AS "7 DIV 2"    run: 3 err=<nil> (translate err <nil>)
    dialect_test.go:60: SELECT MOD(7, 2)         -> SELECT MOD(7, 2)                                           run: 1 err=<nil> (translate err <nil>)
```

The translation leaves `MOD` where it stands, while the `DIV` on the line below it is rewritten.

## Expected behavior

`a MOD b` must be translated to `a % b`, which is the same operation with the same result and the same precedence: MySQL gives `%`, `MOD` and `DIV` one precedence level, and SQLite gives `%` the level of `*` and `/`, so replacing the word with the operator does not change how the expression groups. `dialect/doc.go` puts this in the first of the three classes it names, "Known incompatibilities that have a SQLite equivalent are rewritten", and the third class, pass-through, is for constructs this package does not know about. The package does know about this one: `MOD` is listed beside `DIV` in `operandExpectingKeywords` in `dialect/label.go`, so the alias logic already treats it as an operator while the rewrite passes never see it.

Note while fixing that MySQL's `MOD` and SQLite's `%` agree on the sign of the result — both take the sign of the dividend, so `-7 MOD 2` is `-1` — which is why the translation is a token replacement rather than a helper call.

## Cause (guess)

This is a guess: `mysql.go` has a `divPass` implementing `a DIV b -> CAST(a / b AS INTEGER)` and no pass for `MOD`, so the word reaches SQLite, which has no such operator. The fix is a pass beside `divPass` that replaces a `MOD` word token standing between two operands with a `%` operator token. It has to leave the function call alone: a `MOD` immediately followed by `(` is `MOD(a, b)`, which already works, so the pass keys on the word being in operator position rather than on the word alone. An identifier named `mod` is the case to be careful about, and quoting already separates it — a backtick-quoted `` `mod` `` is an identifier token rather than a word by the time the passes run.

## Tests to add

- In the MySQL translation tests, cases pinning `SELECT 7 MOD 2`, `SELECT a MOD b FROM t`, `SELECT a MOD b AS m FROM t` and `SELECT * FROM t WHERE a MOD b = 1` to the `%` translation, and an execution test asserting `7 MOD 2` is 1 and `-7 MOD 2` is -1, which is the sign rule that makes the replacement sound.
- The case that must not be rewritten: `SELECT MOD(7, 2)` keeps its function spelling and still answers 1, and a column or alias named `mod` (`SELECT mod FROM t`, `SELECT a AS mod FROM t`) is untouched.
- A precedence case, since the point of the replacement is that grouping does not change: `SELECT 1 + 7 MOD 2` and `SELECT 7 MOD 2 * 3` must answer what MySQL answers, 2 and 3.
- While in there, the two dialects' sibling gaps are worth a look and a decision, either a rewrite or `ErrUnsupportedSyntax` rather than the engine's message: GoogleSQL's `DIV(x, y)` fails with `no such function: DIV`, and the two-argument `TRUNC(x, n)` that both PostgreSQL and GoogleSQL define fails with `wrong number of arguments to function TRUNC()`.

### PR DESCRIPTION
**Pull request by nao1215**
Two defects found by probing the dialect translation with differential testing: each translated query was run beside the SQLite query that means the same thing, and the disagreements were then checked against what the source engine answers.

## What changed

A cast of a value the integer type has no room for now answers what the dialect answers. The conversion rounded and then converted to int64 with no range check, and Go leaves that conversion implementation-defined for a value outside the range, so `CAST(1e30 AS SIGNED)`, `(1e30)::bigint` and `CAST(1e30 AS INT64)` all returned `-9223372036854775808`: a positive number read back as the most negative one, silently, and not even the same way on every platform this project builds for. MySQL clamps to the bound of the type, which is the answer it gives with a warning; PostgreSQL and GoogleSQL raise `ErrInvalidCast`, which `SAFE_CAST` turns into NULL. NaN takes the same two answers. The bound is 2^63 rather than the largest integer, because no float64 holds that one, while -2^63 is exact and converts.

MySQL's `MOD` operator is translated rather than handed to SQLite. `SELECT 7 MOD 2` came back as `near "2": syntax error` — the engine message this package exists to keep callers away from — while `DIV`, the same family of operator, was already rewritten. `MOD` is SQLite's `%`: same operation, same precedence, same sign rule, so the translation is a token replacement rather than a helper call. `MOD(a, b)` already worked and is left alone, and a call is told from an operator the way MySQL's own parser tells them apart, by whether the parenthesis follows the name with nothing between.

## How it is kept fixed

The cast range is pinned from both sides: the clamp and the two rejections are table cases beside the existing cast semantics, and the values just inside the range, the largest and smallest integers, and the bound itself are pinned so a range check cannot start rejecting what fits. `roundForDialect` also has a direct test, because SQLite has no way to carry a NaN into a query — a NaN real comes back as NULL — so that branch can only be reached from Go.

`MOD` is pinned as a translation and as an answer: the translation cases cover the operator between names, between literals, in a WHERE clause, with either operand parenthesized, and inside a call's arguments, and the execution cases cover the sign rule (`-7 MOD 2` is -1), the precedence (`1 + 7 MOD 2` is 2, `7 MOD 2 * 3` is 3), and the spellings that must not change — the function call, a quoted column named `mod`, and a quoted alias.

## Verification

`make test`, `make lint`, cross-builds for windows/amd64 and darwin/arm64, and sqly built and tested against this branch through a local replace. A query summing 20,000 casts, at `-count=5`: median 3.545ms before and 3.187ms after, with allocations unchanged, so the range check costs nothing measurable.

Closes #389
Closes #390

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved integer casting for out-of-range values, NaN, infinity, and oversized numeric strings across MySQL, PostgreSQL, and GoogleSQL.
  - MySQL now clamps overflow values, while PostgreSQL and GoogleSQL report casting errors or return `NULL` with `SAFE_CAST`.
  - Corrected MySQL `DIV` precedence and associativity in mixed arithmetic expressions.
  - Added support for translating MySQL infix `MOD` expressions while preserving `MOD(...)` function calls.
- **Tests**
  - Expanded coverage for casting edge cases and MySQL arithmetic translation behavior.

### CODE REVIEW

File: dialect/cast.go [hunk 1]
```diff
@@ -231,27 +231,67 @@ func castToInt(d Dialect, v driver.Value, strict bool) (driver.Value, error) {
...
-		return roundForDialect(d, x), nil
+		return roundForDialect(d, x, strict)
 	}
 	s, _ := toString(v)
 	if n, err := strconv.ParseInt(strings.TrimSpace(s), 10, 64); err == nil {
 		return n, nil
 	}
 	if f, err := strconv.ParseFloat(strings.TrimSpace(s), 64); err == nil {
-		return roundForDialect(d, f), nil
+		return roundForDialect(d, f, strict)
```
[the remark this answers is not in the document]
Comment (nao1215) [reply dialect/cast.go]: Fixed in 2f328cb. ParseFloat is now accepted when it returns strconv.ErrRange, so the infinity it hands back reaches the range check and is clamped for MySQL and rejected for the strict dialects. TestCastStringPastTheFloatRange covers a 400-digit string in all three dialects, which the 20-digit case did not reach.

File: dialect/mysql.go [hunk 2]
```diff
@@ -42,6 +43,9 @@ func rewriteMySQL(tokens []token) ([]token, error) {
 	if err != nil {
 		return nil, err
 	}
+	// M-24: MOD is DIV's sibling, and SQLite's "%" is the same operation at the
+	// same precedence, so this one is a token replacement rather than a call.
+	out = modPass(out)
```
[the remark this answers is not in the document]
Comment (nao1215) [reply dialect/mysql.go]: Confirmed and fixed in fe4f345, and the underlying defect is filed as #392. It was not only MOD: on main, 8 * 5 DIV 2 answered 16 where MySQL answers 20, and 100 / 5 DIV 2 answered 50 where MySQL answers 10, because divPass took only the primary expression before the operator. The left operand now walks back over the operators of that precedence level and is parenthesized in the output so the later division pass cannot regroup it, and MOD is rewritten before DIV so a MOD to the left is already an operator token. Execution cases for 8 MOD 5 DIV 2, 8 * 5 DIV 2, 100 / 5 DIV 2 and 8 % 5 DIV 2 are in TestOperatorSemantics, beside the cases that must not change (2 + 8 DIV 2, 8 DIV 2 DIV 2, ABS(-8) DIV 2).

File: dialect/cast.go [hunk 3]
```diff
@@ -231,27 +231,71 @@ func castToInt(d Dialect, v driver.Value, strict bool) (driver.Value, error) {
...
 		return n, nil
 	}
-	if f, err := strconv.ParseFloat(strings.TrimSpace(s), 64); err == nil {
-		return roundForDialect(d, f), nil
+	// ParseFloat answers a well-formed number too large for a float64 with an
+	// infinity and ErrRange. That is a value, not a parse failure: reading it as
+	// one sent a 400-digit number down the numeric-prefix path below, where it
+	// came back as 0 instead of as the bound of the type.
+	if f, err := strconv.ParseFloat(strings.TrimSpace(s), 64); err == nil || errors.Is(err, strconv.ErrRange) {
+		return roundForDialect(d, f, strict)
```
[the remark this answers is not in the document]
Comment (nao1215) [reply dialect/cast.go]: Fixed in 1a07909. A well-formed integer that ParseInt reports as out of range is now answered from the integer parse, before any float is involved, so the -2^63-1 case no longer arrives at the range check as -2^63. Regression cases are in cast_test.go for all three dialects: MySQL clamps '-9223372036854775809' to the lower bound and '9223372036854775808' to the upper one, PostgreSQL and GoogleSQL reject both, SAFE_CAST returns NULL, and '-9223372036854775808' still converts unchanged.
